lib/puppet/configurer.rb

```diff
@@ -126,6 +126,45 @@ def convert_catalog(result, duration, facts, options = {})
     catalog
   end
 
+  def check_fact_name_length(name)
+    if name.length > Puppet[:max_factnamelength]
+      Puppet.warning _("Fact '%{name}' exceeds the name length limit") % { name: name }
+    end
+  end
+
+  def check_fact_values_length(keys, values)
+    if values.is_a?(Hash) || values.is_a?(Array)
+      values.each do |k,v|
+        value = v || k
+        if value.is_a?(Hash) || value.is_a?(Array)
+          check_fact_values_length(value)
+        else
+          if v.to_s.length > Puppet[:max_factvaluelength]
+            Puppet.warning _("Fact '%{k}' with value '%{v}' exceeds the value length limit") % { k: k, v: v }
+          end
```

Comment:
It's a little confusing/surprising that `v` is nil when `values` is an array. Maybe unpack that logic so the Array and Hash cases are handled separately?